aten/src/ATen/native/Normalization.cpp

```diff
@@ -54,6 +54,60 @@ struct Var {
   }
 };
 
+/// A fast path for CPU inference when all tensors are contiguous.
+template<typename scalar_t>
+void batch_norm_cpu_inference_contiguous(Tensor* output, const Tensor& input,
+    const Tensor& weight /* optional */, const Tensor& bias /* optional */,
+    const Tensor& mean, const Tensor& variance, double eps) {
+  int64_t n_batch = input.size(0);
+  int64_t n_channel = input.size(1);
+  int64_t image_size = input.numel() / n_batch / n_channel;
+
+  scalar_t* output_data = output->data<scalar_t>();
+  const scalar_t* input_data = input.data<scalar_t>();
+  const scalar_t* weight_data = weight.defined() ? weight.data<scalar_t>() : nullptr;
+  const scalar_t* bias_data = bias.defined() ? bias.data<scalar_t>() : nullptr;
+  const scalar_t* mean_data = mean.data<scalar_t>();
+  const scalar_t* var_data = variance.data<scalar_t>();
+
+  /// Collect the linear and constant terms regarding the input.
+  /// output(n, c, h, w)
+  ///     = (input(n, c, h, w) - mean(c)) / sqrt(var(c) + eps) * weight(c)
+  ///         + bias(c)
+  ///     = input(n, c, h, w) * inv_var(c) * weight(c) +
+  ///         - mean(c) * inv_var(c) * weight(c) + bias(c),
+  /// where inv_var(c) = 1 / sqrt(var(c) + eps).
+  /// So the linear term, alpha(c) = inv_var(c) * weight(c),
+  ///   the constant term beta(c) = bias(c) - mean(c) * inv_var(c) * weight(c)
+  Tensor alpha = at::empty_like(mean);
+  Tensor beta = at::empty_like(mean);
+  scalar_t* alpha_data = alpha.data<scalar_t>();
+  scalar_t* beta_data = beta.data<scalar_t>();
+  for (int64_t c = 0; c < n_channel; c++) {
+    scalar_t inv_var = 1 / std::sqrt(var_data[c] + eps);
```

Comment:
The RHS of this expression is computed in double precision when scalar_t=float, then it is being narrowed to float on assignment. Is that intended?

Reply:
The LLVM vectorizer does this with that:

```
        vpslld  $31, %ymm5, %ymm5
        vpsrad  $31, %ymm5, %ymm5
        xorl    %r10d, %r10d
        movq    -64(%rbp), %r11         ## 8-byte Reload
        movq    -56(%rbp), %r14         ## 8-byte Reload
        .p2align        4, 0x90
LBB0_33:                                ## =>This Inner Loop Header: Depth=1
        vcvtps2pd       (%rdi,%rbx), %ymm6
        vcvtps2pd       16(%rdi,%rbx), %ymm7
        vaddpd  %ymm7, %ymm2, %ymm7
        vaddpd  %ymm6, %ymm2, %ymm6
        vsqrtpd %ymm6, %ymm6
        vsqrtpd %ymm7, %ymm7
        vdivpd  %ymm7, %ymm3, %ymm7
        vdivpd  %ymm6, %ymm3, %ymm6
        vcvtpd2ps       %ymm6, %xmm6
        vcvtpd2ps       %ymm7, %xmm7
        vinsertf128     $1, %xmm7, %ymm6, %ymm6
        vmovups (%rsi,%rbx), %ymm7
        vmaskmovps      (%rdx,%rbx), %ymm4, %ymm8
        vpandn  %ymm8, %ymm5, %ymm8
        vmulps  %ymm6, %ymm7, %ymm9
        vmovups %ymm9, (%r14,%rbx)
        vmulps  (%rcx,%rbx), %ymm6, %ymm6
        vmulps  %ymm6, %ymm7, %ymm6
        vsubps  %ymm6, %ymm8, %ymm6
        vmovups %ymm6, (%r11,%rbx)
        vcvtps2pd       32(%rdi,%rbx), %ymm6
        vcvtps2pd       48(%rdi,%rbx), %ymm7
        vaddpd  %ymm7, %ymm2, %ymm7
        vaddpd  %ymm6, %ymm2, %ymm6
        vsqrtpd %ymm6, %ymm6
        vsqrtpd %ymm7, %ymm7
        vdivpd  %ymm7, %ymm3, %ymm7
        vdivpd  %ymm6, %ymm3, %ymm6
        vcvtpd2ps       %ymm6, %xmm6
        vcvtpd2ps       %ymm7, %xmm7
        vinsertf128     $1, %xmm7, %ymm6, %ymm6
        vmovups 32(%rsi,%rbx), %ymm7
        vmaskmovps      32(%rdx,%rbx), %ymm4, %ymm8
        vpandn  %ymm8, %ymm5, %ymm8
        vmulps  %ymm6, %ymm7, %ymm9
        vmovups %ymm9, 32(%r14,%rbx)
        vmulps  32(%rcx,%rbx), %ymm6, %ymm6
        vmulps  %ymm6, %ymm7, %ymm6
        vsubps  %ymm6, %ymm8, %ymm6
        vmovups %ymm6, 32(%r11,%rbx)
        addq    $16, %r10
        addq    $64, %rbx
        addq    $2, %rax
        jne     LBB0_33
## %bb.34:
        testq   %r8, %r8
        je      LBB0_36
```

Reply:
I pulled it into godbolt here: https://godbolt.org/z/OdsS-W

Reply:
I think we should figure out what flags we shuold use when analyzing this stuff. iirc the configuration is:

1) FBCode: -mavx
2) OSS build (default): (none)
3) I build with `-march=native -mavx -mavx2` on my mac 

Do you know what flags are used in the distributed binaries?

Reply:
I just confirmed that the OSS build with default settings emits SSE instructions

Reply:
The distributed binaries use runtime dispatch and are compiled with -mavx and -mavx2 -mfma if I remember this correctly.

Reply:
OSS only compiles code with `-mavx` (and `-mavx2`) if it's in the `cpu/` subdirectory. This isn't in that directory so you'll only get standard x86-64 instructions (including SSE)

Reply:
Regarding "The RHS of this statement is computed in double precision when scalar_t=float, then it is being narrowed to float on assignment. Is that intended?". 

It was intended for better accuracy because often the epsilon was very small. And it would have no performance implication if (C << total_size). However, in degenerate cases where (N==1 && image_size==1), meaning (C==total_size), this is quite slow. So I change the default to casting epsilon in this case.